src/vcpkg/registries.cpp

```diff
@@ -713,40 +711,48 @@ namespace
     ExpectedL<std::unique_ptr<RegistryEntry>> BuiltinFilesRegistry::get_port_entry(StringView port_name) const
     {
         auto port_directory = m_builtin_ports_directory / port_name;
-        if (m_fs.exists(port_directory, IgnoreErrors{}))
+        const auto& maybe_maybe_scf = get_scf(port_name, port_directory);
+        const auto maybe_scf = maybe_maybe_scf.get();
+        if (!maybe_scf)
         {
-            const auto& found_scf = get_scf(port_directory);
-            if (auto scf = found_scf.get())
-            {
-                if (scf->core_paragraph->name == port_name)
-                {
-                    return std::make_unique<BuiltinPortTreeRegistryEntry>(
-                        scf->core_paragraph->name, port_directory, scf->to_version());
-                }
+            return maybe_maybe_scf.error();
+        }
 
-                return msg::format_error(msgUnexpectedPortName,
-                                         msg::expected = scf->core_paragraph->name,
-                                         msg::actual = port_name,
-                                         msg::path = port_directory);
-            }
+        auto scf = maybe_scf->get();
+        if (!scf)
+        {
+            return std::unique_ptr<RegistryEntry>();
+        }
 
-            return found_scf.error();
+        if (scf->core_paragraph->name == port_name)
+        {
+            return std::make_unique<BuiltinPortTreeRegistryEntry>(
+                scf->core_paragraph->name, port_directory, scf->to_version());
         }
 
-        return nullptr;
+        return msg::format_error(msgUnexpectedPortName,
+                                 msg::expected = scf->core_paragraph->name,
+                                 msg::actual = port_name,
+                                 msg::path = port_directory);
     }
 
-    ExpectedL<Version> BuiltinFilesRegistry::get_baseline_version(StringView port_name) const
+    ExpectedL<Optional<Version>> BuiltinFilesRegistry::get_baseline_version(StringView port_name) const
     {
         // if a baseline is not specified, use the ports directory version
-        auto port_path = m_builtin_ports_directory / port_name;
-        const auto& maybe_scf = get_scf(port_path);
-        if (auto pscf = maybe_scf.get())
+        const auto& maybe_maybe_scf = get_scf(port_name, m_builtin_ports_directory / port_name);
+        auto maybe_scf = maybe_maybe_scf.get();
+        if (!maybe_scf)
         {
-            return pscf->to_version();
+            return maybe_maybe_scf.error();
         }
 
-        return maybe_scf.error();
+        auto scf = maybe_scf->get();
+        if (!scf)
+        {
+            return Optional<Version>();
+        }
+
+        return scf->to_version();
```

Comment:
suggestion:
```cpp
return get_scf(port_name, m_builtin_ports_directory / port_name)
  .map([](const std::unique_ptr<SourceControlFile>& scf)
    {
      return !scf ? Optional<Version>() : scf->to_version();
    });
```

Reply:
The suggestion introduces an extra move, as `map` forces the temporary `Optional<Version>{scf->to_version()}` to be materialized.

I think that move might be worth it if we think this is substantially clearer but I'm not sure it is...

`then` avoids the extra move but forces the type to be repeated again as the lambda's return type.